From 69aa216239d51897f005d44605ed2b88e63d477b Mon Sep 17 00:00:00 2001
From: Xulin Sun <xulin.sun@windriver.com>
Date: Mon, 14 Dec 2020 13:25:20 +0800
Subject: [PATCH] pfeng-napi: stop phylink when releasing pfeng port

pfeng_phylink_stop() should be called when releasing pfeng port.
Otherwise there will be call trace like below:

[ 3162.602085] pfeng 46000000.pfe pfe2: Set TX clock to 125M
[ 3162.607649] called from state RUNNING
[ 3162.607802] WARNING: CPU: 1 PID: 7055 at drivers/net/phy/phy.c:916 phy_start+0x54/0xbc
[ 3162.607809] Modules linked in: 8021q pfeng(O) sch_fq_codel openvswitch nsh nf_conncount nf_nat nf_conntrack nf_defrag_ipv6 nf_defrag_ipv4
[ 3162.607850] CPU: 1 PID: 7055 Comm: ifconfig Tainted: G W O 5.4.80-yocto-standard #1
[ 3162.607855] Hardware name: Freescale S32G274 (DT)
[ 3162.607863] pstate: 20000005 (nzCv daif -PAN -UAO)
[ 3162.607871] pc : phy_start+0x54/0xbc
[ 3162.607879] lr : phy_start+0x54/0xbc
[ 3162.607884] sp : ffffffc0174539c0
[ 3162.607888] x29: ffffffc0174539c0 x28: 0000000000000000
[ 3162.607897] x27: ffffff8869dae400 x26: 0000000000000041
[ 3162.607906] x25: ffffff88705b9260 x24: 0000000000000000
[ 3162.607916] x23: 0000000000001002 x22: ffffffc00900efd0
[ 3162.613217] pfeng 46000000.pfe pfe2: Link is Up - 1Gbps/Full - flow control off
[ 3162.619129] x21: ffffff885a798040 x20: ffffff8870769800
[ 3162.619138] x19: ffffff8870769000 x18: 0000000000000000
[ 3162.619146] x17: 0000000000000000 x16: 0000000000000000
[ 3162.619155] x15: 000000000000b6ce x14: 000000000000ba7e
[ 3162.619163] x13: 000000000001d018 x12: ffffffc0123a2000
[ 3162.619172] x11: ffffffc011723000 x10: ffffffc0123a2330
[ 3162.619180] x9 : 0000000000000000 x8 : ffffffc011702000
[ 3162.619191] x7 : 0000000000000000 x6 : ffffffc011701000
[ 3162.730925] x5 : ffffffc011701558 x4 : 0000000000000000
[ 3162.736219] x3 : 0000000000000000 x2 : ffffff885a798040
[ 3162.741514] x1 : 6a91d8dff9f64500 x0 : 0000000000000000
[ 3162.746810] Call trace:
[ 3162.749245] phy_start+0x54/0xbc
[ 3162.752455] phylink_start+0xb4/0x27c
[ 3162.756160] pfeng_phylink_start+0x24/0x34 [pfeng]
[ 3162.760907] pfeng_logif_open+0x158/0x220 [pfeng]
[ 3162.765566] __dev_open+0x108/0x1b0
[ 3162.769033] __dev_change_flags+0x1b4/0x220
[ 3162.773200] dev_change_flags+0x40/0x80
[ 3162.777021] devinet_ioctl+0x674/0x6e4
[ 3162.780752] inet_ioctl+0x1dc/0x260
[ 3162.784223] sock_do_ioctl+0x58/0x230
[ 3162.787867] sock_ioctl+0x1f4/0x494
[ 3162.791343] do_vfs_ioctl+0x344/0x420
[ 3162.794985] ksys_ioctl+0x88/0xc0
[ 3162.798285] __arm64_sys_ioctl+0x2c/0xd0
[ 3162.802193] el0_svc_common.constprop.0+0x78/0x13c
[ 3162.806963] el0_svc_handler+0x54/0xc0
[ 3162.810699] el0_svc+0x8/0x208
[ 3162.813733] irq event stamp: 0
[ 3162.816775] hardirqs last enabled at (0): [<0000000000000000>] 0x0
[ 3162.823026] hardirqs last disabled at (0): [<ffffffc0100b23a4>] copy_process+0x2e4/0x1000
[ 3162.831183] softirqs last enabled at (0): [<ffffffc0100b23a4>] copy_process+0x2e4/0x1000
[ 3162.839341] softirqs last disabled at (0): [<0000000000000000>] 0x0

Upstream-Status: Inappropriate [Upstream with latest version, has replaced these
series code with other API functions.]

Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 sw/linux-pfeng/pfeng-napi.c | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/sw/linux-pfeng/pfeng-napi.c b/sw/linux-pfeng/pfeng-napi.c
index ac83d93..c3c806c 100644
--- a/sw/linux-pfeng/pfeng-napi.c
+++ b/sw/linux-pfeng/pfeng-napi.c
@@ -203,15 +203,9 @@ static int pfeng_logif_release(struct net_device *netdev)
 
 	netdev_info(netdev, "%s\n", __func__);
 
-#if LINUX_VERSION_CODE < KERNEL_VERSION(5,4,0)
-	/*
-	 * Note: phylink_stop() causes backtraces on 5.4
-	 *       but worked correctly on 4.19
-	 */
 	/* stop phylink */
 	if (ndev->phylink)
 		pfeng_phylink_stop(ndev);
-#endif
 
 	/* stop napi */
 	netif_tx_stop_queue(netdev_get_tx_queue(netdev, 0));
-- 
2.17.1

